first draft of readme - need to add in simple code-along style test(s) still (#1)
learn-co-curriculum/working-with-typescript: README.md

# Working with TypeScript Code Along

## Learning Goals

- Successfully set up a local environment to run TypeScript
- Configure and understand the tsconfig.json file
- Practice compiling with `tsc`

## Introduction

TypeScript is a very useful and powerful tool we can use to optimize our code by
checking for syntax errors before runtime, giving tips on how to fix bugs before
they happen and more, all by adding in types to our regular JavaScript code!
Though we can start using TypeScript right out of the box, it does recognize
that every application and developer is different, and gives us some room for
flexibility. We get access to this flexibility/customization through the
ts.config file that we will keep in the root of our application, where we can
change default rules, and add some rules of our own!

## Environment Setup

- fork and clone this lab, then open in your text editor

> - install ts globally

Reviewer: Should this be locally?

Author: oh those are just my notes i forgot to delete whoopsies

> You should already have Node Installed on your machine, but if you haven't,

Reviewer: Suggest changing to "... but if you don't"

Author: ```suggestion
You should already have Node Installed on your machine, but if you don't,
```

please do so following the [Node.js Installation Docs][]. If you would like a
more detailed walk through, please check out the following repositories:

- [Installing NVM & Node.js on MacOS][]

> - [Installing NVM & Node.js on WSL2][]

Reviewer: If you use this markdown for links, you need to have the link defined somewhere in the file using a colon rather than parentheses, eg:

`[Installing NVM & Node.js on MacOS ]: <url>`

And you also need the `[Installing NVM & Node.js on MacOS][]` syntax everywhere you want the link to appear

Author: oh woops i thought i added those lolol

Author: oh wait nvm i see what you mean. ty!

While we can install TypeScript globally, it's generally best practice to do so
on a project-by-project basis to avoid different versioning across machines.
Let's install TypeScript to this project by running the following code in your
terminal _(Remember to cd into the correct project directory!)_:

```zsh
$ npm install typescript --save-dev
```

Once TypeScript is done installing into our project, you should see some
node_modules files appear into our file directory. Our next step is adding in
our `tsconfig.json` file.

## The tsconfig.json file

So what _is_ the tsconfig.json file? The tsconfig file is used to help us
configure and compile our code! This file is _always_ kept in the root of our
project, and is used to identify the compiler rules we want to enforce, as well
as specify the root files of the project. The rules the compiler enforces are
completely customizable and up to you to determine based on your project needs.

> There's unlimited ways that you can setup your tsconfig.json file, and a lot of

Reviewer: Should be set up rather than setup (grammar wonk - sorry)

Author: ```suggestion
There's unlimited ways that you can set up your tsconfig.json file, and a lot of
```

different properties you can add to it.

Let's create our tsconfig.json file in the root of our project directory by
running:

```zsh
$ touch tsconfig.json
```

Awesome. Now let's fill in our properties. Were going to start with a

> `"compilerOptions"` key with an object as it's value. The `"compilerOptions"`

Reviewer: should be its rather than it's (also punctuation wonk)

Author: ```suggestion
`"compilerOptions"` key with an object as its value. The `"compilerOptions"`
```

object property is what we will use to declare the rules we want to enforce in
our project. Inside of the `"compilerOptions"` object, we're going to add 3
properties.

> Let's start by adding the property`"module"`, and setting it's value to

Reviewer: Here too with the "its"

Author: ```suggestion
Let's start by adding the property `"module"`, and setting its value to
```

browsers we could set our value to include the year (e.g. `es2022`), or even
`esnext`.

Our final property to add is`"strictNullChecks"`, which makes sure no variable
has the value of `null` or `undefined` unless we explicitly assign them.

Let's add one more property to our json object, `"include": ["**/*.ts"]`. The
`include` property tells our compiler what files to check our rules against. Our
value, `["**/*.ts"]`, tells the compiler to run these rules against every file
in our project that has the extension `.ts`.

You should now have an object that looks something like this:

```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es6",
    "strictNullChecks": true
  },
  "include": ["**/*.ts"]
}
```

For this exercise, we're working with a very simple `tsconfig.json` file, but
it's definitely worth taking a look at the [TypeScript Docs - tsconfig][] for
more compiler options and properties to add to your projects in the future!

## Compiling with `tsc`

- run tsc to see the errors
- then fix the errors
- then run tsc to see there are no more errors
- run node index.js to run our code
- submit code

> Now that we have our `tsconfig.json` setup, we're able to compile our code using

Reviewer: Here too with the "set up"

Author: ```suggestion
Now that we have our `tsconfig.json` set up, we're able to compile our code using
```

the `tsc` command! There's a ton of different compiler options for us (check out
[TypeScript Docs - Compiler Options][] to see them all!) to use... we can use

> the command by adding a specific file to it like `npx tsc index.ts`, or we can

Reviewer: Suggest moving the parenthetical info after "to use"

Author: Does this look better? 

> Now that we have our `tsconfig.json` setup, we're able to compile our code using
> the `tsc` command! There's a ton of different compiler options for us to use  (check out
> [TypeScript Docs - Compiler Options][] to see them all!), we can use
> the command by adding a specific file to it like `npx tsc index.ts`, or we can
